<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>biogram - analysis of biological sequences using n-grams — biogram-package • biogram</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="biogram - analysis of biological sequences using n-grams — biogram-package" />
<meta property="og:description" content="biogram package is a toolbox for the analysis of
nucleic acid and protein sequences using n-grams. Possible applications include
motif discovery, feature selection, clustering, and classification." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biogram</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.6.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/overview.html">biogram package</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/michbur/biogram">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>biogram - analysis of biological sequences using n-grams</h1>
    <small class="dont-index">Source: <a href='https://github.com/michbur/biogram/blob/master/R/biogram.R'><code>R/biogram.R</code></a></small>
    <div class="hidden name"><code>biogram-package.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>biogram</code> package is a toolbox for the analysis of
nucleic acid and protein sequences using n-grams. Possible applications include
motif discovery, feature selection, clustering, and classification.</p>
    </div>



    <h2 class="hasAnchor" id="n-grams"><a class="anchor" href="#n-grams"></a>n-grams</h2>

    

<p>n-grams (k-tuples) are sets of <code>n</code> characters derived from the input sequence(s).
They may form continuous sub-sequences or be discontinuous. For example, from the 
sequence of nucleotides <code>AATA</code> one can extract the following continuous 
2-grams (bigrams): <code>AA</code>, <code>AT</code> and <code>TA</code>. Moreover, there are two
possible bigrams separated by a single space: <code>A_T</code> and <code>A_A</code>, and one
bigram separated by two spaces: <code>A__A</code>.</p>
<p>Another important n-gram parameter is its position. Instead of just counting n-grams,
one may want to count how many n-grams occur at a given position in multiple (e.g. related)
sequences. For example, in the sequences <code>AATA</code> and <code>AACA</code> there is only one
bigram at position 1: <code>AA</code>, but there are two bigrams at position two: <code>AT</code> and
<code>AC</code>. The following notation is used for position-specific n-grams: <code>1_AA</code>,
<code>2_AT</code>, <code>2_AC</code>.</p>
<p>In the <code>biogram</code> package, the <code><a href='count_ngrams.html'>count_ngrams</a></code> function is used for
counting and extracting n-grams. Using the <code>d</code> argument the user can specify the
distance between elements of the n-grams. The <code>pos</code> argument can be used to enable
position specificity.</p>
    <h2 class="hasAnchor" id="n-gram-data-dimensionality"><a class="anchor" href="#n-gram-data-dimensionality"></a>n-gram data dimensionality</h2>

    

<p>We note that n-grams suffer from the curse of dimensionality. For example, for a peptide
of length 6 \(20^{n}\) n-grams and \(6 \times 20^{n}\) positioned n-grams are possible.
Data sets of such an enormous size are hard to manage and analyze in R.</p>
<p>The <code>biogram</code> package deals with both of the abovementioned problems. It uses 
innate properties of the n-gram data which usually can be represented by sparse
matrices. Data storage is done using functionalities from the <code>slam</code> package. To ease 
the selection of significant features, <code>biogram</code> provides the user with QuiPT, 
a very fast permutation test for binary data (see <code><a href='test_features.html'>test_features</a></code>).</p>
<p>Another way of reducing dimensionality is the aggregation of sequence residues into more 
general groups. For example, all positively-charged amino acids may be aggregated into
one group. This action can be performed using the <code><a href='degenerate.html'>degenerate</a></code> function.</p>
<p>Encoding of amino acids can easu sequence analysis, but multidimensional 
objects as the aggregations of amino acids are not easily comparable. We introduced the 
encoding distance, a measure defining the distance between encodings. It can be computed 
using the <code><a href='calc_ed.html'>calc_ed</a></code> function.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># use data set from package</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>human_cleave</span>)
<span class='co'># first nine columns represent subsequent nine amino acids from cleavage sites</span>
<span class='co'># degenerate the sequence to reduce the dimensionality of the problem</span>
<span class='co'># (use five groups instead of 20 amino acids)</span>
<span class='no'>deg_seqs</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='degenerate.html'>degenerate</a></span>(<span class='no'>human_cleave</span>[, <span class='fl'>1L</span>:<span class='fl'>9</span>],
                      <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='kw'>`a`</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>1</span>, <span class='fl'>6</span>, <span class='fl'>8</span>, <span class='fl'>10</span>, <span class='fl'>11</span>, <span class='fl'>18</span>),
                           <span class='kw'>`b`</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>2</span>, <span class='fl'>13</span>, <span class='fl'>14</span>, <span class='fl'>16</span>, <span class='fl'>17</span>),
                           <span class='kw'>`c`</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>5</span>, <span class='fl'>19</span>, <span class='fl'>20</span>),
                           <span class='kw'>`d`</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>7</span>, <span class='fl'>9</span>, <span class='fl'>12</span>, <span class='fl'>15</span>),
                           <span class='st'>'e'</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>3</span>, <span class='fl'>4</span>)))
<span class='co'># EXAMPLE 1 - extract significant trigrams</span>
<span class='co'># extract trigrams</span>
<span class='no'>trigrams</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='count_ngrams.html'>count_ngrams</a></span>(<span class='no'>deg_seqs</span>, <span class='fl'>3</span>, <span class='no'>letters</span>[<span class='fl'>1L</span>:<span class='fl'>5</span>], <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='co'># select features that differ between the two target groups using QuiPT</span>
<span class='no'>test1</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='test_features.html'>test_features</a></span>(<span class='no'>human_cleave</span>[, <span class='st'>"tar"</span>], <span class='no'>trigrams</span>)
<span class='co'># see a summary of the results</span>
<span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span>(<span class='no'>test1</span>)</div><div class='output co'>#&gt; Total number of features: 690 
#&gt; Number of significant features: 70 
#&gt; Criterion used: Information Gain 
#&gt; Feature test: QuiPT 
#&gt; p-values adjustment method: BH </div><div class='input'><span class='co'># aggregate features in groups based on their p-value</span>
<span class='no'>gr</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cut.html'>cut</a></span>(<span class='no'>test1</span>)
<span class='co'># get position map of the most significant n-grams</span>
<span class='fu'><a href='position_ngrams.html'>position_ngrams</a></span>(<span class='no'>gr</span><span class='kw'>[[</span><span class='fl'>1</span>]])</div><div class='output co'>#&gt; $`1`
#&gt; [1] a_0 a_0 b_0
#&gt; Levels: a_0 b_0 c_0 d_0 e_0
#&gt; 
#&gt; $`2`
#&gt; [1] a_0 a_0 a_0 b_0 b_0 b_0
#&gt; Levels: a_0 b_0 c_0 d_0 e_0
#&gt; 
#&gt; $`3`
#&gt;  [1] a_0 a_0 a_0 a_0 a_0 a_0 a_0 a_0 a_0 a_0 b_0 d_0
#&gt; Levels: a_0 b_0 c_0 d_0 e_0
#&gt; 
#&gt; $`4`
#&gt;  [1] a_0 a_0 a_0 a_0 a_0 b_0 b_0 c_0 d_0 e_0
#&gt; Levels: a_0 b_0 c_0 d_0 e_0
#&gt; 
#&gt; $`5`
#&gt; [1] a_0 a_0 a_0 a_0 a_0 a_0 a_0 a_0
#&gt; Levels: a_0 b_0 c_0 d_0 e_0
#&gt; 
#&gt; $`6`
#&gt; [1] a_0 a_0
#&gt; Levels: a_0 b_0 c_0 d_0 e_0
#&gt; 
#&gt; $`7`
#&gt; [1] b_0
#&gt; Levels: a_0 b_0 c_0 d_0 e_0
#&gt; </div><div class='input'><span class='co'># transform the most significant n-grams to more readable form</span>
<span class='fu'><a href='decode_ngrams.html'>decode_ngrams</a></span>(<span class='no'>gr</span><span class='kw'>[[</span><span class='fl'>1</span>]])</div><div class='output co'>#&gt; 1_a.a.a_0.0 2_a.a.a_0.0 3_a.a.a_0.0 4_a.a.a_0.0 1_b.a.a_0.0 2_b.a.a_0.0 
#&gt;       "aaa"       "aaa"       "aaa"       "aaa"       "baa"       "baa" 
#&gt; 3_b.a.a_0.0 1_a.b.a_0.0 3_a.b.a_0.0 3_a.c.a_0.0 3_a.d.a_0.0 3_a.e.a_0.0 
#&gt;       "baa"       "aba"       "aba"       "aca"       "ada"       "aea" 
#&gt; 5_a.a.b_0.0 2_b.d.b_0.0 
#&gt;       "aab"       "bdb" </div><div class='input'>
<span class='co'># EXAMPLE 2 - search for specific n-grams</span>
<span class='co'># the n-grams of the interest are a_a (a-gap-a) and e_e (e-gap-e) on the</span>
<span class='co'># 3rd and 4th position</span>
<span class='co'># firstly code n-grams in biogram notation and add position information</span>
<span class='no'>coded</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='code_ngrams.html'>code_ngrams</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"a_a"</span>, <span class='st'>"c_c"</span>))
<span class='co'># add position information</span>
<span class='no'>coded</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span>(<span class='st'>"3_"</span>, <span class='no'>coded</span>), <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span>(<span class='st'>"4_"</span>, <span class='no'>coded</span>))
<span class='co'># count only the features of the interest</span>
<span class='no'>bigrams</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='count_specified.html'>count_specified</a></span>(<span class='no'>deg_seqs</span>, <span class='no'>coded</span>)
<span class='co'># test which of the features of the interest is significant</span>
<span class='no'>test2</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='test_features.html'>test_features</a></span>(<span class='no'>human_cleave</span>[, <span class='st'>"tar"</span>], <span class='no'>bigrams</span>)
<span class='fu'><a href='https://rdrr.io/r/base/cut.html'>cut</a></span>(<span class='no'>test2</span>)</div><div class='output co'>#&gt; $`[0,0.0001]`
#&gt; [1] "3_a.a_1"
#&gt; 
#&gt; $`(0.0001,0.01]`
#&gt; [1] "3_c.c_1" "4_a.a_1"
#&gt; 
#&gt; $`(0.01,0.05]`
#&gt; character(0)
#&gt; 
#&gt; $`(0.05,1]`
#&gt; [1] "4_c.c_1"
#&gt; </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#n-grams">n-grams</a></li>
      <li><a href="#n-gram-data-dimensionality">n-gram data dimensionality</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Michal Burdukiewicz, Piotr Sobczyk, Chris Lauber</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Michal Burdukiewicz, Piotr Sobczyk, Chris Lauber, Dominik Rafacz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


